File: homework/81333/golang-highlighter/src/Data/Highlighter.hs
{-# LANGUAGE OverloadedStrings #-}

module Data.Highlighter
  ( tokensToHtml
  ) where

import qualified Data.Text       as T
import qualified Data.Text.IO    as TI

import           Data.Token
import           Data.TokenTypes

initialHtml :: T.Text
initialHtml =
  T.unlines
    [ "<html>"
    , "<head>"
    , "<style>"
    , ".keyword { color: orange; }"
    , ".comment { color: grey; }"
    , ".preident { color: red; }"
    , ".operator { color: black; }"
    , ".ident { color: blue; }"
    , ".string { color: green; }"
    , ".number { color: purple; }"
    , "</style>"
    , "</head>"
    , "<body>"
    ]

toHtml :: Token -> T.Text
toHtml (Token {tokenType = Space _, lexeme = lex, location = (line, col)}) = lex
toHtml (Token {tokenType = Comment _, lexeme = lex, location = (line, col)}) =
  "<span class=comment>" <> lex <> "</span>"
toHtml (Token {tokenType = Keyword _, lexeme = lex, location = (line, col)}) =
  "<span class=keyword>" <> lex <> "</span>"
toHtml (Token {tokenType = PredeclIdent _, lexeme = lex, location = (line, col)}) =
  "<span class=preident>" <> lex <> "</span>"
toHtml (Token {tokenType = Operator _, lexeme = lex, location = (line, col)}) =
  "<span class=operator>" <> lex <> "</span>"
toHtml (Token {tokenType = Ident _, lexeme = lex, location = (line, col)}) =
  "<span class=ident>" <> lex <> "</span>"
toHtml (Token { tokenType = Literal (StringLiteral _)
              , lexeme = lex
              , location = (line, col)
              }) = "<span class=string>" <> lex <> "</span>"
toHtml (Token { tokenType = Literal (Number _)
              , lexeme = lex
              , location = (line, col)
              }) = "<span class=number>" <> lex <> "</span>"

tokensToHtml :: [Token] -> T.Text
tokensToHtml tokens =
  initialHtml <> T.concat (map toHtml tokens) <> "</body><br></html>"

File: homework/81352/ts2html-hw1/highlighter.cpp
#include <sstream>
#include <iostream> 
#include <unordered_set>
#include "highlighter.h"

Highlighter::Highlighter(std::vector<Token> tokens): tokens(tokens){}

std::string Highlighter::highlight() {
    std::ostringstream result;
    result << "<!doctype><html><head>";
    result << "<style>.keyword {color:red;} \
                      .operator {color:#0e7c31; font-weight: bold} \
                      .number {color:#d119b2;} .string {color:blue;} \
                      .comment {color:grey; font-style: italic;} \
                      .line_number {opacity: 0.5;} \
                      .decorator {color:#1a1f63; font-style: italic;} \
                      .user_defined_type {color:#105109} \
               </style></head><body><pre>";
    if (!this->tokens.empty()) {
        result << "<span class=\"line_number\">1. </span>";
    }
    int line_number = 2;
    for (Token token: this->tokens) { 
        if (token.getTokenType() == TokenType::NEWLINE) {
            result << "\n<span class=\"line_number\">" << line_number++ << ". </span>";
        } else {
            result << "<span class=\"" << token.getTypeName() << "\">" << token.getTextValue() << "</span>";
        }
    }
    result << "</pre></body></html>";
    return result.str();
}
